src/app/util/MatterCallbacks.h

```diff
@@ -14,32 +13,61 @@
  *    See the License for the specific language governing permissions and
  *    limitations under the License.
  */
-
-// THIS FILE IS GENERATED BY ZAP
-
 #pragma once
 
 #include <access/SubjectDescriptor.h>
 #include <app/ConcreteAttributePath.h>
 #include <app/ConcreteCommandPath.h>
 
-void MatterPreAttributeReadCallback(const chip::app::ConcreteAttributePath & attributePath);
-void MatterPostAttributeReadCallback(const chip::app::ConcreteAttributePath & attributePath);
-void MatterPreAttributeWriteCallback(const chip::app::ConcreteAttributePath & attributePath);
-void MatterPostAttributeWriteCallback(const chip::app::ConcreteAttributePath & attributePath);
+namespace chip {
 
-/** @brief Matter Pre Command Received
- *
- * This callback is called once the message has been determined to be a command, and
- * before the command is dispatched to the receiver.
- */
-CHIP_ERROR MatterPreCommandReceivedCallback(const chip::app::ConcreteCommandPath & commandPath,
-                                            const chip::Access::SubjectDescriptor & subjectDescriptor);
+/// Allows for application hooks for processing attribute and command operations
+class DataModelCallbacks
+{
+public:
+    enum class OperationType
+    {
+        Read,
+        Write
+    };
 
-/** @brief Matter Post Command Received
- *
- * This callback is called once the message has been determined to be a command, but
- * after it being dispatched to the receiver.
- */
-void MatterPostCommandReceivedCallback(const chip::app::ConcreteCommandPath & commandPath,
-                                       const chip::Access::SubjectDescriptor & subjectDescriptor);
+    enum class OperationOrder
+    {
+        Pre,
+        Post
+    };
+
+    virtual ~DataModelCallbacks() = default;
+
+    /// This callback is called on attribute operations:
+    /// - for reads and writes
+    /// - both before and after attribute read/writes
+    ///
+    /// NOTE: PostRead is only called on read success.
+    virtual void AttributeOperation(OperationType operation, OperationOrder order, const chip::app::ConcreteAttributePath & path) {}
+
+    /// This callback is called once for every command dispatch, before the dispatch is actually
+    /// done towards the receiver.
+    ///
+    /// This method is called once for every CommandDataIB (i.e. it may be called several times
+    /// in the case of batch invoke, where a single `InvokeRequestMessage` may contain several
+    /// CommandDataIB entries).
+    ///
+    /// Returning an error here will prevent the command to be dispatched further
+    virtual CHIP_ERROR PreCommandReceived(const chip::app::ConcreteCommandPath & commandPath,
+                                          const chip::Access::SubjectDescriptor & subjectDescriptor)
+    {
+        return CHIP_NO_ERROR;
+    }
+
+    /// This callback is called once the message has been determined to be a command, but
+    /// after it being dispatched to the receiver.
```

Comment:
This comment is wrong too.

Yes, I know you just copied it.  Filing a followup issue on the comment fix would have been fine, since the broken comment was not being introduced by this PR.  But we should be minimally tracking things we know are broken.

Thank you for fixing the comment on the PreCommandReceived bits.